app/gui/src/controller/searcher/component.rs

```diff
@@ -222,34 +233,34 @@ pub struct ModuleGroups {
 /// [`builder::List`].
 #[derive(Clone, CloneRef, Debug, Default)]
 pub struct List {
-    all_components:        Rc<Vec<Component>>,
-    top_modules:           group::AlphabeticalList,
-    top_modules_flattened: group::AlphabeticalList,
-    module_groups:         Rc<HashMap<Id, ModuleGroups>>,
-    filtered:              Rc<Cell<bool>>,
+    all_components:             Rc<Vec<Component>>,
+    top_module_sections:        Rc<Vec<Section>>,
+    top_module_section_indices: Rc<HashMap<ImString, usize>>,
+    module_groups:              Rc<HashMap<Id, ModuleGroups>>,
+    filtered:                   Rc<Cell<bool>>,
     /// Components to display in the "Local Scope" section of the [Component
     /// Browser](crate::controller::Searcher).
-    pub local_scope:       Group,
+    pub local_scope:            Group,
     /// Groups of components to display in the "Favorites Data Science Tools" section of the
     /// [Component Browser](crate::controller::Searcher).
-    pub favorites:         group::List,
+    pub favorites:              group::List,
 }
 
 impl List {
     /// Return the list of top modules, which should be displayed in Component Browser.
     ///
     /// If the list is filtered, all top modules will be flattened.
-    pub fn top_modules(&self) -> &group::AlphabeticalList {
+    pub fn top_modules(&self) -> Vec<group::AlphabeticalList> {
         if self.filtered.get() {
-            &self.top_modules_flattened
+            self.top_module_sections.iter().map(|s| s.modules_flattened.clone()).collect()
         } else {
-            &self.top_modules
+            self.top_module_sections.iter().map(|s| s.modules.clone()).collect()
```

Comment:
I wonder if we could return `impl Iterator<Item=group::AlphabeticalList>`, and put the if inside map lambda. We replace then additional allocation with more "if"s, but I think the former is still a better (in performance terms) option. What do you think?

And, IIRC, the modules could be cloned_ref, not cloned (and in getter the performance of clones is important.

Reply:
That makes sense. I figured we always collect the result to a vector in the caller anyway, but returning an iterator makes for a better interface. I've applied the suggested changes.